### PR TITLE
esp32s3: Add RTC support

### PR DESCRIPTION
**Pull request by acassis**
## Summary
Add RTC support to ESP32S3
## Impact
Only ESP32-S3
## Testing
ESP32S3-Devkit

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by acassis** [comment]
@xiaoxiang781216 could you please merge?

**Comment by xiaoxiang781216** [comment]
But the ci is broken.

**Comment by acassis** [comment]
hmm, strange this LVGL error, let me check, let me restart the failing CI job

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 1]
```diff
@@ -0,0 +1,2774 @@
...
+#define RTC_POWERUP_CYCLES       OTHER_BLOCKS_POWERUP
+#define RTC_WAIT_CYCLES          OTHER_BLOCKS_WAIT
+
+#define DG_WRAP_POWERUP_CYCLES   OTHER_BLOCKS_POWERUP
+#define DG_WRAP_WAIT_CYCLES      OTHER_BLOCKS_WAIT
+
+#define RTC_MEM_POWERUP_CYCLES   OTHER_BLOCKS_POWERUP
+#define RTC_MEM_WAIT_CYCLES      OTHER_BLOCKS_WAIT
+
+#define RTC_CNTL_PLL_BUF_WAIT_SLP   2
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:81]:
```suggestion
#define OTHER_BLOCKS_WAIT           1

#define ROM_RAM_POWERUP_CYCLES      OTHER_BLOCKS_POWERUP
#define ROM_RAM_WAIT_CYCLES         OTHER_BLOCKS_WAIT

#define WIFI_POWERUP_CYCLES         OTHER_BLOCKS_POWERUP
#define WIFI_WAIT_CYCLES            OTHER_BLOCKS_WAIT

#define RTC_POWERUP_CYCLES          OTHER_BLOCKS_POWERUP
#define RTC_WAIT_CYCLES             OTHER_BLOCKS_WAIT

#define DG_WRAP_POWERUP_CYCLES      OTHER_BLOCKS_POWERUP
#define DG_WRAP_WAIT_CYCLES         OTHER_BLOCKS_WAIT

#define RTC_MEM_POWERUP_CYCLES      OTHER_BLOCKS_POWERUP
#define RTC_MEM_WAIT_CYCLES         OTHER_BLOCKS_WAIT

#define RTC_CNTL_PLL_BUF_WAIT_SLP   2
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 2]
```diff
@@ -0,0 +1,2774 @@
...
+    {
+      slow_freq = esp32s3_rtc_clk_slow_freq_get();
+      if (slow_freq == RTC_SLOW_FREQ_32K_XTAL)
+        {
+          cal_clk = RTC_CAL_32K_XTAL;
+        }
+      else if (slow_freq == RTC_SLOW_FREQ_8MD256)
+            {
+              cal_clk = RTC_CAL_8MD256;
+            }
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:539]:
```suggestion
      else if (slow_freq == RTC_SLOW_FREQ_8MD256)
        {
          cal_clk = RTC_CAL_8MD256;
        }
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 3]
```diff
@@ -0,0 +1,2774 @@
...
+  else if (cal_clk == RTC_CAL_8MD256 || slow_freq == RTC_SLOW_FREQ_8MD256)
+    {
+      expected_freq = RTC_FAST_CLK_FREQ_APPROX / 256;
+    }
+  else
+    {
+      expected_freq = 150000; /* 150k internal oscillator */
+    }
+
+  us_time_estimate = (uint32_t) (((uint64_t) slowclk_cycles) *
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:582]:
```suggestion
  us_time_estimate = (uint32_t) (((uint64_t)slowclk_cycles) *
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 4]
```diff
@@ -0,0 +1,2774 @@
...
+      esp32s3_rtc_clk_slow_freq_set(rtc_slow_freq);
+      if (SLOW_CLK_CAL_CYCLES > 0)
+        {
+          /* 32k XTAL oscillator has some frequency drift at startup. Improve
+           * calibration routine to wait until the frequency is stable.
+           */
+
+          cal_val = esp32s3_rtc_clk_cal(RTC_CAL_RTC_MUX,
+                                      SLOW_CLK_CAL_CYCLES);
+        }
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:872]:
```suggestion
          cal_val = esp32s3_rtc_clk_cal(RTC_CAL_RTC_MUX,
                                        SLOW_CLK_CAL_CYCLES);
        }
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 5]
```diff
@@ -0,0 +1,2774 @@
...
+      else
+        {
+          cal_dividend = (1ULL << RTC_CLK_CAL_FRACT) * 1000000ULL;
+          cal_val = (uint32_t) (cal_dividend /
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:876]:
```suggestion
          cal_val = (uint32_t)(cal_dividend /
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 6]
```diff
@@ -0,0 +1,2774 @@
...
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void IRAM_ATTR esp32s3_rtc_clk_cpu_freq_to_8m(void)
+{
+  int origin_soc_clk = REG_GET_FIELD(SYSTEM_SYSCLK_CONF_REG,
+                                        SYSTEM_SOC_CLK_SEL);
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:904]:
```suggestion
  int origin_soc_clk = REG_GET_FIELD(SYSTEM_SYSCLK_CONF_REG,
                                     SYSTEM_SOC_CLK_SEL);
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 7]
```diff
@@ -0,0 +1,2774 @@
...
+  int origin_div_cnt = REG_GET_FIELD(SYSTEM_SYSCLK_CONF_REG,
+                                        SYSTEM_PRE_DIV_CNT);
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:906]:
```suggestion
  int origin_div_cnt = REG_GET_FIELD(SYSTEM_SYSCLK_CONF_REG,
                                     SYSTEM_PRE_DIV_CNT);
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 8]
```diff
@@ -0,0 +1,2774 @@
...
+  ets_update_cpu_frequency(8);
+
+  REGI2C_WRITE_MASK(I2C_DIG_REG, I2C_DIG_REG_EXT_DIG_DREG,
+                                          DIG_DBIAS_XTAL);
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:910]:
```suggestion
  REGI2C_WRITE_MASK(I2C_DIG_REG, I2C_DIG_REG_EXT_DIG_DREG,
                    DIG_DBIAS_XTAL);
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 9]
```diff
@@ -0,0 +1,2774 @@
...
+ *
+ ****************************************************************************/
+
+static void IRAM_ATTR esp32s3_rtc_clk_cpu_freq_to_pll_mhz(
+                                             int cpu_freq_mhz)
+{
+  int origin_soc_clk = REG_GET_FIELD(SYSTEM_SYSCLK_CONF_REG,
+                                        SYSTEM_SOC_CLK_SEL);
+  int origin_cpuperiod_sel = REG_GET_FIELD(SYSTEM_CPU_PER_CONF_REG,
+                                              SYSTEM_CPUPERIOD_SEL);
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:945]: ditto

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 10]
```diff
@@ -0,0 +1,2774 @@
...
+  enum esp32s3_rtc_slow_freq_e rtc_slow_freq_8md256 =
+                                         RTC_SLOW_FREQ_8MD256;
+  enum esp32s3_rtc_cal_sel_e cal_clk = RTC_CAL_RTC_MUX;
+  if (slow_clk_freq == rtc_slow_freq_x32k)
+    {
+      cal_clk = RTC_CAL_32K_XTAL;
+    }
+  else if (slow_clk_freq == rtc_slow_freq_8md256)
+    {
+      cal_clk  = RTC_CAL_8MD256;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:1076]:
```suggestion
      cal_clk = RTC_CAL_8MD256;
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 11]
```diff
@@ -0,0 +1,2774 @@
...
+ ****************************************************************************/
+
+uint64_t IRAM_ATTR esp32s3_rtc_time_get(void)
+{
+  uint64_t rtc_time;
+
+  modifyreg32(RTC_CNTL_RTC_TIME_UPDATE_REG, 0, RTC_CNTL_RTC_TIME_UPDATE);
+
+  rtc_time = getreg32(RTC_CNTL_TIME0_REG);
+  rtc_time |= ((uint64_t) getreg32(RTC_CNTL_TIME1_REG)) << 32;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:1686]:
```suggestion
  rtc_time |= ((uint64_t)getreg32(RTC_CNTL_TIME1_REG)) << 32;
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 12]
```diff
@@ -0,0 +1,2774 @@
...
+ *   time_in_us      - Time interval in microseconds
+ *   slow_clk_period - Period of slow clock in microseconds
+ *
+ * Returned Value:
+ *   Number of slow clock cycles
+ *
+ ****************************************************************************/
+
+uint64_t IRAM_ATTR esp32s3_rtc_time_us_to_slowclk(uint64_t time_in_us,
+                                                uint32_t period)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:1707]:
```suggestion
uint64_t IRAM_ATTR esp32s3_rtc_time_us_to_slowclk(uint64_t time_in_us,
                                                  uint32_t period)
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.c [hunk 13]
```diff
@@ -0,0 +1,2774 @@
...
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void IRAM_ATTR esp32s3_rtc_cpu_freq_set_xtal(void)
+{
+  int freq_mhz = (int) esp32s3_rtc_clk_xtal_freq_get();
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.c:1850]:
```suggestion
  int freq_mhz = (int)esp32s3_rtc_clk_xtal_freq_get();
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.h [hunk 14]
```diff
@@ -0,0 +1,769 @@
...
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#ifndef __ARCH_XTENSA_SRC_ESP32_ESP32_RTC_H
+#define __ARCH_XTENSA_SRC_ESP32_ESP32_RTC_H
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.h:26]:
```suggestion
#ifndef __ARCH_XTENSA_SRC_ESP32S3_ESP32S3_RTC_H
#define __ARCH_XTENSA_SRC_ESP32S3_ESP32S3_RTC_H
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.h [hunk 15]
```diff
@@ -0,0 +1,769 @@
...
+};
+
+/* Structure used to pass parameters to set an alarm */
+
+struct alm_setalarm_s
+{
+  int               as_id;     /* enum alm_id_e */
+  struct timespec   as_time;   /* Alarm expiration time */
+  alm_callback_t    as_cb;     /* Callback (if non-NULL) */
+  void          *as_arg;       /* Argument for callback */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.h:173]:
```suggestion
  alm_callback_t    as_cb;     /* Callback (if non-NULL) */
  void              *as_arg;   /* Argument for callback */
```

File: arch/xtensa/src/esp32s3/esp32s3_rtc.h [hunk 16]
```diff
@@ -0,0 +1,769 @@
...
+
+#endif /* CONFIG_RTC_DRIVER */
+
+#ifdef __cplusplus
+}
+#endif
+#undef EXTERN
+
+#endif /* __ASSEMBLY__ */
+#endif /* __ARCH_XTENSA_SRC_ESP32_ESP32_RTC_H */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtc.h:769]:
```suggestion
#endif /* __ARCH_XTENSA_SRC_ESP32S3_ESP32S3_RTC_H */
```

File: arch/xtensa/src/esp32s3/esp32s3_rtcheap.c [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+ * Name: esp32s3_rtcheap_initialize
+ *
+ * Description:
+ *   Initialize the RTC heap.
+ *
+ ****************************************************************************/
+
+void esp32s3_rtcheap_initialize(void)
+{
+  void  *start;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtcheap.c:53]:
```suggestion
  void *start;
```

File: arch/xtensa/src/esp32s3/esp32s3_rtcheap.c [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+  size_t size;
+
+  /* These values come from the linker scripts. Check boards/xtensa/esp32. */
+
+  extern uint8_t _srtcheap[];
+  extern uint8_t _ertcheap[];
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_rtcheap.c:59]: can we move this out of the function?

**Comment by acassis** [comment]
Fixed! Thank you very much @pkarashchenko !!!

**Pull request merged**